src/events.rs

```diff
@@ -339,35 +339,118 @@ pub enum EventsDecodingError {
-//                 )))],
-//                 &mut &input[..],
-//                 &mut output,
-//                 &mut errors,
-//             )
-//             .unwrap();
-//
-//         assert_eq!(output, vec![1, 0]);
-//     }
-// }
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        Config,
+        DefaultConfig,
+        Phase,
+    };
+    use frame_metadata::{
+        v14::{
+            ExtrinsicMetadata,
+            PalletEventMetadata,
+            PalletMetadata,
+            RuntimeMetadataLastVersion,
+        },
+        RuntimeMetadataPrefixed,
+    };
+    use scale_info::{
+        meta_type,
+        TypeInfo,
+    };
+    use std::convert::TryFrom;
+
+    #[derive(Encode)]
+    pub struct EventRecord<E: Encode> {
+        phase: Phase,
+        pallet_index: u8,
+        event: E,
+        topics: Vec<<DefaultConfig as Config>::Hash>,
+    }
+
+    fn event_record<E: Encode>(pallet_index: u8, event: E) -> EventRecord<E> {
+        EventRecord {
+            phase: Phase::Finalization,
+            pallet_index,
+            event,
+            topics: vec![],
+        }
+    }
+
+    fn pallet_metadata<E: TypeInfo + 'static>(pallet_index: u8) -> PalletMetadata {
+        let event = PalletEventMetadata {
+            ty: meta_type::<E>(),
+        };
+        PalletMetadata {
+            name: "Test",
+            storage: None,
+            calls: None,
+            event: Some(event),
+            constants: vec![],
+            error: None,
+            index: pallet_index,
+        }
+    }
+
+    fn init_decoder(pallets: Vec<PalletMetadata>) -> EventsDecoder<DefaultConfig> {
+        let extrinsic = ExtrinsicMetadata {
+            ty: meta_type::<()>(),
+            version: 0,
+            signed_extensions: vec![],
+        };
+        let v14 = RuntimeMetadataLastVersion::new(pallets, extrinsic, meta_type::<()>());
+        let runtime_metadata: RuntimeMetadataPrefixed = v14.into();
+        let metadata = Metadata::try_from(runtime_metadata).unwrap();
+        EventsDecoder::<DefaultConfig>::new(metadata)
+    }
+
+    #[test]
+    fn decode_single_event() {
+        #[derive(Clone, Encode, TypeInfo)]
+        enum Event {
+            A(u8),
+        }
+
+        let pallet_index = 0;
+        let pallet = pallet_metadata::<Event>(pallet_index);
+        let decoder = init_decoder(vec![pallet]);
+
+        let event = Event::A(1);
+        let encoded_event = event.encode();
+        let event_records = vec![event_record(pallet_index, event)];
+
+        let mut input = Vec::new();
+        event_records.encode_to(&mut input);
+
+        let events = decoder.decode_events(&mut &input[..]).unwrap();
+
+        assert_eq!(events[0].1.variant_index, encoded_event[0]);
+        assert_eq!(events[0].1.data.0, encoded_event[1..]);
+    }
+
+    #[test]
+    fn compact_event_field() {
+        #[derive(Clone, Encode, TypeInfo)]
+        enum Event {
+            A(#[codec(compact)] u32),
```

Comment:
This all looks good; is it worth adding another test to check that compact encoded structs containing a single compact-encoded field are ok too (it looks like the code handles them too)?

Reply:
Good idea, added some more tests. Can be more too still, but at least we now have a template for testing this component.